Add a detailed ReadMe (#1)
sdsc-ordes/zarr-linked-data: README.md

This project came into being tied to the [Cat+ initiative from EPF domain in Switzerland](https://swisscatplus.ch). From a fully automated chemistry lab, numerous samples will be processed under different parameters by a variety of machines. The idea is to have the output data of this lab, and even more so the associated metadata, to be queryable and retrievable. On the long run, the idea is also to allow external sources to contribute their data to the system as well, following metadata standards. 

The *main functionalities* that were implemented in this project were: 
- Retrieve (["consolidate"](https://zarr.readthedocs.io/en/stable/tutorial.html#consolidating-metadata) in zarr jargon) all the metadata from the numerous arrays of data
- From a metadata Universal Resource Identifier [URI](https://www.w3.org/wiki/URI), retrieve the associated dataset in the [Zarr store](https://zarr.readthedocs.io/en/stable/api/storage.html)

# B. Getting Started

## Locally (without k8)

Set-up you [poetry project](https://python-poetry.org/docs/#installation) using `poetry install` in the same folder as the pyproject.toml

OR 

> Install the requirements with `pip install -r requirements.txt` (these are the requirements for the kubernetes set-up so you will have some useless packages for local dev)

Reviewer: ```suggestion
Install the requirements with `pip install -r requirements.txt` (these are the requirements for the kubernetes set-up so you will install packages which are not needed for local dev)
```
**nitpick(non-blocking):** wording

## With kubernetes

You will be using the manifests in the `manifest` folder.

### Set up a cluster

Install [minikube](https://minikube.sigs.k8s.io/docs/start/) (really for simple prototyping) or [K3S](https://k3s.io) (already production oriented). 
This project used minikube.

In this cluster create a namespace where you will be deploying all your other components. In our project it is called `argo`.

### Set up an S3 storage

> We will need a storage for our fake data, for metaflow flow code packages, and for Argo artifacts (more below). You can use `argo-minio.yaml` [in the manifests folder](manifests/argo-minio.yaml) or set-up your own [via the MinIO documentation (for prototyping)](https://min.io/docs/minio/kubernetes/upstream/index.html)

Reviewer: **question:** What is the command needed to use `argo-mini.yaml`?

Author: I have added the kubectl command

Create a secret for allowing other services to access MinIO and its bucket storage: 
```
kubectl create secret generic argo-artifacts
--from-literal=accesskey=XXXXXXX
--from-literal=secretkey=XXXXXXXXXXXXXXX
-n argo
```

### Set up Argo for automation

Installation: 
1. [This installation was used](https://argoproj.github.io/argo-workflows/quick-start/): `kubectl apply -n argo -f https://github.com/argoproj/argo-workflows/releases/download/v3.4.8/install.yaml`
2. Argo deployment is launched with: `kubectl -n argo port-forward deployment/argo-server 2746:2746` 

Artifact Repository: 
Argo needs an artifact repository (storage for flows to run). Here we will use MinIO that we previously installed. You can use the `artifact-repositories.yaml` [in the manifests folder](manifests/artifact-repositories.yaml)
(The secret for minio we created before comes in here). You can set it up with: `kubectl apply -f artifact-repositories.yaml`

> Give Argo admin roles on cluster to access MinIO: : `kubectl create rolebinding argo-default-admin --clusterrole=admin --serviceaccount=argo:default --namespace=argo` (we also had to repete this rolebinding creating for argo and argo-server service accounts in the argo namespace. This rolebinding may have to be revised in a production environment where an admin role could be problematic.

Reviewer: ```suggestion
Give Argo admin roles on cluster to access MinIO: : `kubectl create rolebinding argo-default-admin --clusterrole=admin --serviceaccount=argo:default --namespace=argo` (we also had to repeat this rolebinding creating for argo and argo-server service accounts in the argo namespace. This role-binding may have to be revised in a production environment where an admin role could be problematic.
```
**fix:** typos

### Set up an IDE (optional)

> If for some reason, you need/want an IDE deployed (if for example you are working on a server). Install via Helm [a vscode service](https://artifacthub.io/packages/helm/inseefrlab/vscode) then access by port-forwarding `kubectl -n argo port-forward deployment/my-vscode 35547:8080`. You will probably have to give admin rights to this VSCode (service account: `my-vscode` over MinIO (as done for Argo)). Finally, you will also have to add Metaflow configuration variables to the ConfigMap of VSCode, for putting metaflow code packages in MinIO. 

Reviewer: ```suggestion
If you want an IDE deployed (e.g. working on a server). Install [a vscode service](https://artifacthub.io/packages/helm/inseefrlab/vscode) via Helm, then access by port-forwarding `kubectl -n argo port-forward deployment/my-vscode 35547:8080`. You will probably have to give admin rights to this VSCode (service account: `my-vscode` over MinIO (as done for Argo)). Finally, you will also have to add Metaflow configuration variables to the ConfigMap of VSCode, for putting metaflow code packages in MinIO. 
```
**suggestion(non-blocking):** more concise wording

![configmap-my-vscode-configmaps3metaflow](https://github.com/SDSC-ORD/zarr-linked-data/assets/22447169/498a1306-5289-4527-9cfa-22e947d555e3)

Then you just install the requirements with `pip install -r requirements.txt` and you're set to go! 

# C. Example/Usage

## Fake Zarr Data

> You may not have data under the zarr format yet. You can define some metadata instance objects (such as in `zarr_linked_data/data/original_data.jsonld`), then define your hierarchy levels, data level and other parameters for the creation of a random Zarr test store using `zarr_linked_data/fake_data_flow.py`.

Reviewer: **comment(non-blocking):** This (versioned) data makes the repo quite large (350MB), which can be cumbersome / slow things down. Perhaps we could later consider moving it to a data repository (zenodo?). It would also require rewriting the git-history to entirely remove the files from the index, though (with git commands or [BFG](https://rtyley.github.io/bfg-repo-cleaner/))

Author: agreed, I had to put it in the repo to get it on the collaborator's server. not the best but it was the moment's quick fix :)

Author: I will add it to the Roadmap

- Explore Zarr group and attributes functionalities with `python zarr_linked_data/local_dev/zarr_explore.py`
- Fake data creation pipeline with: `python zarr_linked_data/fake_data_flow.py`
- Metadata consolidation Metaflow pipeline with: `python zarr_linked_data/local_dev/consolidate_metadata_flow.py run`
- Retrieve URI Metaflow pipeline with: `python zarr_linked_data/local_dev/retrieval_flow.py run`

## Scripts on kubernetes

- Fake data creation pipeline with: `python zarr_linked_data/fake_data_flow.py` (same as for locally)
- Metadata consolidation Metaflow pipeline with: `python zarr_linked_data/consolidate_metadata_flow.py run`
- Retrieve URI Metaflow pipeline with: `python zarr_linked_data/retrieval_flow.py run`

You can check the correct run of the metaflow flows with the command specified in the script. Then you will need to send them to Argo workflows. [Using Metaflow to create Argo DAGs](https://docs.metaflow.org/production/scheduling-metaflow-flows/scheduling-with-argo-workflows): `python zarr_linked_data/retrieval_flow.py --with retry argo-workflows create` (same for `consolidate_metadata_flow`)

# Roadmap

- Retrieval flow will be converted to a FastAPI instead

> - A flow `metadata update` will be added: it will transform the consolidated metadata and add it to a graph database (such as GraphDB or ApacheJenaFuseki): local development of [this flow is on an annex branch here](https://github.com/SDSC-ORD/zarr-linked-data/blob/metadata_extractor/zarr_linked_data/extract_all_metadata.py)

Reviewer: **note:** awesome! I'm very interested :grinning:

Author: If I can, I will make it for local_dev in my refactor next week

Author: Actually I will first do a PR for local dev and another one for this extra flow
